nistats/reporting/glm_reporter.py

```diff
@@ -0,0 +1,913 @@
+    
+    Parameters
+    ----------
+    contrasts: Dict[str, np.array or str]
+        Contrast information, as a dict
+          {'contrast_title_1, contrast_info_1/title_1, ...}
+    
+    design_matrices: List[pd.Dataframe]
+        Design matrices computed in the model.
+
+    Returns
+    -------
+    contrast_plots: Dict[str, svg img]
+        Dict of contrast title and svg image data url
+        for corresponding contrast plot.
+    """
+    all_contrasts_plots = {}
+    contrast_template_path = os.path.join(html_template_root_path,
+                                          'contrast_template.html'
+                                          )
+    with open(contrast_template_path) as html_template_obj:
+        contrast_template_text = html_template_obj.read()
+    
+    for design_matrix in design_matrices:
+        for contrast_name, contrast_data in contrasts.items():
+            contrast_text_ = string.Template(contrast_template_text)
+            contrast_plot = plot_contrast_matrix(contrast_data, design_matrix)
+            contrast_plot.set_xlabel(contrast_name)
+            contrast_plot.figure.set_figheight(2)
+            contrast_plot.figure.set_tight_layout(True)
+            url_contrast_plot_svg = plot_to_svg(contrast_plot)
+            contrasts_for_subsitution = {
+                'contrast_plot': url_contrast_plot_svg,
+                'contrast_name': contrast_name,
+                }
+            contrast_text_ = contrast_text_.safe_substitute(
+                    contrasts_for_subsitution
+                    )
+            all_contrasts_plots[contrast_name] = contrast_text_
+    return all_contrasts_plots
+
+
+def _make_headings(contrasts, title, model):
+    """ Creates report page title, heading & sub-heading
+     using title text or contrast names.
+    Accepts contrasts and user supplied title string or
+    contrasts and user supplied 3 element list or tuple.
+    
+    If title is not in (None, 'auto'),
+    page title == heading,
+    model type == sub-heading
+    
+    Parameters
+    ----------
+    contrasts: Dict[str, np.array or str]
+        Contrast information, as a dict in the form
+            {'contrast_title_1': contrast_info_1/title_1, ...}
+        Contrast titles are used in page title and secondary heading
+        if `title` is not 'auto' or None.
+    
+    title: String or List/Tuple with 3 elements
+        User supplied text for HTML Page title and primary heading.
+        Or 3 element List/Tuple for Title Heading, sub-heading resp.
+        Overrides title auto-generation.
+    
+    model: FirstLevelModel or SecondLevelModel
+        The model, passed in to determine its type
+        to be used in page title & headings.
+    Returns
+    -------
+    (HTML page title, heading, sub-heading): Tuple[str, str, str]
+        If title is user-supplied, then subheading is empty string.
+    """
+    if isinstance(title, (tuple, list)) and len(title) == 3:
```

Comment:
get exactly one string or None from the user; if not none, display it as a top-level heading.
generate the subtitles -- first or second level model etc.